examples/dither.dx

```diff
@@ -0,0 +1,157 @@
+Width = Fin 300
+key = new_key 1234
+img_ = for i:Height. for j:Width. n_to_f ((ordinal i) + (ordinal j))
+img = img_ / (n_to_f (size Height + size Width))
+
+:t img
+> ((Fin 100) => (Fin 300) => Float32)
+:html matshow img
+> <html output>
+
+' # Round 1: Threshold dithering
+
+def dither_threshold {h w} (img: h=>w=>Float) : (h=>w=>Bool) =
+  for i j. img.i.j > 0.5
+
+:html bmatshow $ dither_threshold img
+> <html output>
+
+' # Round 2: Noise dithering
+
+def dither_noise {h w} (k:Key) (img: h=>w=>Float) : (h=>w=>Bool) =
+  for i j. img.i.j > rand (ixkey k (i, j))
+
+:html bmatshow $ dither_noise key img
+> <html output>
+
+' # Round 3: Ordered dithering
+  Ordered dithering uses a small threshold matrix that is periodically tiled over the target
+  image. We achieve this by wrapping the pixel indices i and j into the range
+  of the small matrix `mat` to retrieve the threshold.
+
+def dither_ordered {mh mw h w} (mat: mh=>mw=>Float) (img: h=>w=>Float) : (h=>w=>Bool) =
+  for i j. img.i.j > mat.(wrap mh (ordinal i)).(wrap mw (ordinal j))
+
+bayer = [[1.0,  9,  3, 11], 
+         [ 13,  5, 15,  7],
+         [  4, 12,  2, 10],
+         [ 16,  8, 14,  6]] / 17.0
+
+' This so-called Bayer matrix leads to the distinctive crosshatch pattern
+  that was frequently used in old video games with limited color palettes.
+
+:html bmatshow $ dither_ordered bayer img
+> <html output>
+
+halftone = [[14.0,  10,  6, 13, 19, 23, 27, 20],
+            [   7,   2,  1,  9, 26, 31, 32, 24],
+            [  11,   3,  4,  5, 22, 30, 29, 28],
+            [  15,   8, 12, 16, 18, 25, 21, 17],
+            [  19,  23, 27, 20, 14, 10,  6, 13],
+            [  26,  31, 32, 24,  7,  2,  1,  9],
+            [  22,  30, 29, 28, 11,  3,  4,  5],
+            [  18,  25, 21, 17, 15,  8, 12, 16]] / 33.0
+
+' Such a "halftone" matrix imitates an [old printing technique for newspapers](https://en.wikipedia.org/wiki/Halftone):
+
+:html bmatshow $ dither_ordered halftone img
+> <html output>
+
+' # Dithering a real image
+
+-- Image Loading Utilities
+
+data Image =
+  MkImage rows:Nat cols:Nat pixels:(Fin rows => Fin cols => Fin 3 => Word8)
+
+def parse_p6 : Parser Image = MkParser \h.
+  -- Loads a raw PPM file in P6 format.
+  -- The header will look something like:
+  --   P6
+  --   220 220    (width, height)
+  --   255        (max color value)
+  -- followed by a flat block of height x width x 3 chars.
+  parse h $ p_char 'P'
+  parse h $ p_char '6'
+  parse h $ parse_any
+  cols = i32_to_n $ parse h $ parse_unsigned_int
+  parse h $ parse_any
+  rows = i32_to_n $ parse h $ parse_unsigned_int
+  parse h $ parse_any
+  colorsize = i32_to_n $ parse h $ parse_unsigned_int
+  parse h $ parse_any
+  pixels = for r:(Fin rows).
+    for c:(Fin cols).
+      for c:(Fin 3).
+        parse h parse_any
+  MkImage rows cols pixels
+
+def load_image_p6 (filename:String) : Maybe Image =
+  image_raw = unsafe_io do read_file filename
+  run_parser_partial image_raw parse_p6
```

Comment:
It might be better to separate those functions into a library in `lib`. Could you please move them there? Perhaps `netpbm.dx`?

Reply:
Sure! We moved the loader to a new `lib/netpbm.dx` now.